File: 981-time-based-key-value-store/981-time-based-key-value-store.cpp
class TimeMap {
public:
    unordered_map<string, vector<pair<int, string>>> mp;
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        mp[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        if (mp.find(key) == mp.end()) {
            return "";
        }
        int l = 0, r = mp[key].size();
        while(l<r){
            int mid = (l+r) >> 1;
            if (timestamp < mp[key][mid].first){
                r = mid;
            }
            else {
                l = mid + 1;
            }
                
        }
        return l > 0 and l<=mp[key].size() ? mp[key][l-1].second : "";
            
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */

File: 981-time-based-key-value-store/981-time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.map = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.map[key].append([timestamp,value])

    def get(self, key: str, timestamp: int) -> str:
        if key in self.map:
            l, r = 0, len(self.map[key])
            while l < r:
                mid = (l + r) >> 1
                if timestamp >= self.map[key][mid][0] :
                    l = mid + 1
                else:
                    r = mid
            return self.map[key][r-1][1] if r != 0 else ""
                
        return "" 
        


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)